creusot-contracts/src/std/option.rs

```diff
@@ -35,63 +35,276 @@ extern_spec! {
+                where
+                    F: FnOnce(T) -> U;
+
+                #[requires(match self {
+                    None => default.precondition(()),
+                    Some(t) => f.precondition((t,)),
+                })]
+                #[ensures(match self {
+                    None => default.postcondition_once((), result),
+                    Some(t) => f.postcondition_once((t,), result),
+                })]
+                fn map_or_else<U, D, F>(self, default: D, f: F) -> U
+                where
+                    D: FnOnce() -> U,
+                    F: FnOnce(T) -> U;
+
+                #[ensures(match self {
+                    None => result == Err(err),
+                    Some(t) => result == Ok(t),
+                })]
+                fn ok_or<E>(self, err: E) -> Result<T, E>;
+
+                #[requires(self == None ==> err.precondition(()))]
+                #[ensures(match self {
+                    None => exists<r: _> result == Err(r) && err.postcondition_once((), r),
+                    Some(t) => result == Ok(t),
+                })]
+                fn ok_or_else<E, F>(self, err: F) -> Result<T, E>
+                where
+                    F: FnOnce() -> E;
+
+                #[ensures(match *self {
+                    None => result == None,
+                    Some(t) => result == Some(&t.deref()),
+                })]
+                fn as_deref(&self) -> Option<&<T as ::std::ops::Deref>::Target>
+                where
+                    T: ::std::ops::Deref;
 
                 #[pure]
                 #[ensures(self == None ==> result == None)]
-                #[ensures(self == None || result == optb)]
+                #[ensures(self == None || (result == optb && self.resolve()))]
                 fn and<U>(self, optb: Option<U>) -> Option<U>;
 
+                #[requires(match self {
+                    None => true,
+                    Some(t) => f.precondition((t,)),
+                })]
+                #[ensures(match self {
+                    None => result == None,
+                    Some(t) => f.postcondition_once((t,), result),
+                })]
+                fn and_then<U, F>(self, f: F) -> Option<U>
+                where
+                    F: FnOnce(T) -> Option<U>;
+
+                #[requires(match self {
+                    None => true,
+                    Some(t) => predicate.precondition((&t,)) &&
+                        // `predicate` returns either true or false, but not both
+                        (predicate.postcondition_once((&t,), true) != predicate.postcondition_once((&t,), false)),
+                })]
+                #[ensures(match self {
+                    None => result == None,
+                    Some(t) => if predicate.postcondition_once((&t,), true) {
+                        result == Some(t)
+                    } else {
+                        result == None && resolve(&t)
+                    },
+                })]
```

Comment:
Shorter spec, which does not requires determinism of the predicate:


```suggestion
                #[ensures(match self {
                    None => result == None,
                    Some(t) =>
                           predicate.postcondition_once((&t,), true) && result == Some(t)
                        || predicate.postcondition_once((&t,), false) && result == None && resolve(&t)
                    },
                })]
```

Reply:
Or even:

```suggestion
                #[ensures(match self {
                    None => result == None,
                    Some(t) => match result {
                        None => predicate.postcondition_once((&t,), false) && resolve(&t),
                        Some(r) => predicate.postcondition_once((&t,), true) && r == t
                    },
                })]
```